File: Fortran/Types/Reduce/works.f90
MODULE subs

  USE iso_c_binding
  USE mpi
  IMPLICIT NONE

  INTEGER, PARAMETER :: nitems = 100

  CONTAINS

  !Simple adder that works for this specific case
  SUBROUTINE adder(newvals, accumvals, nels, datatype)

    INTEGER, DIMENSION(nitems * nels), INTENT(IN) :: newvals
    INTEGER, DIMENSION(nitems * nels), INTENT(INOUT) :: accumvals
    INTEGER, INTENT(IN) :: nels
    INTEGER, INTENT(IN) :: datatype

    accumvals = accumvals + newvals

  END SUBROUTINE adder


  !More complex adder
  SUBROUTINE adder_f2008(newvals, accumvals, nels, datatype)

    TYPE(C_PTR), VALUE :: newvals, accumvals
    INTEGER, INTENT(IN) :: nels
    INTEGER, INTENT(IN) :: datatype
    INTEGER, POINTER, DIMENSION(:) :: nv2, av2
    INTEGER(KIND=MPI_ADDRESS_KIND) :: lb, extents, els, start
    INTEGER :: ierr

    !Find the extents of the datatype
    CALL MPI_Type_get_extent(datatype, lb, extents, ierr)
    !Convert it into numbers of integers (F2008 C_SIZEOF function)
    start = lb / C_SIZEOF(nels)
    els = extents / C_SIZEOF(nels)

    !Behind the scenes newvals and accumvals are C void* pointers
    !Convert them to Fortran Integer pointers (F2003 C_F_POINTER function)
    CALL C_F_POINTER(newvals, nv2, SHAPE=[els])
    CALL C_F_POINTER(accumvals, av2, SHAPE=[els])

    !Add together the elements
    av2 = av2 + nv2

  END SUBROUTINE adder_f2008

END MODULE subs

PROGRAM type

  USE mpi
  USE subs
  IMPLICIT NONE

  INTEGER, PARAMETER :: tag = 100

  INTEGER :: rank
  INTEGER :: nproc
  INTEGER :: ierr
  INTEGER :: contig_type, custom_sum
  INTEGER, DIMENSION(nitems) :: values, values_recv, values_recv_ref

  CALL MPI_Init(ierr)

  CALL MPI_Comm_size(MPI_COMM_WORLD, nproc, ierr)
  CALL MPI_Comm_rank(MPI_COMM_WORLD, rank, ierr)

  values = rank

  !Create the type
  CALL MPI_Type_contiguous(nitems, MPI_INTEGER, contig_type, ierr)

  !Register the type
  CALL MPI_Type_commit(contig_type, ierr)

  !Create the custom addition operator
  CALL MPI_Op_create(adder_f2008, .TRUE., custom_sum, ierr)

  CALL MPI_Reduce(values, values_recv, 1, contig_type, custom_sum, 0, &
      MPI_COMM_WORLD, ierr)

  CALL MPI_Reduce(values, values_recv_ref, nitems, MPI_INTEGER, MPI_SUM, 0, &
      MPI_COMM_WORLD, ierr)

  IF (rank == 0) THEN
    PRINT *,"Sum over all processors = ", SUM(values_recv)
    PRINT *,"Sum over all processors (reference version) = ", &
        SUM(values_recv_ref)
  END IF

  CALL MPI_Finalize(ierr)

END PROGRAM type

File: C/Types/Reduce/works.c
#include <stdio.h>
#include <mpi.h>

#define NITEMS 100

void adder(void* newvals, void* accumvals, int *nels, MPI_Datatype *type)
{
  MPI_Aint lb, extent;
  MPI_Aint cycle, end;
  int *nv2, *av2;

  nv2 = newvals;
  av2 = accumvals;

  MPI_Type_get_extent(*type, &lb, &extent);
  //Convert from bytes and element count to number of integers
  end = extent * (MPI_Aint)*nels / ((MPI_Aint)sizeof(int));
  for (cycle = 0; cycle < end; ++cycle){
    av2[cycle] = av2[cycle] + nv2[cycle];
  }
}

int main(int argc, char** argv)
{

  int rank, recv_rank, nproc, left, right, iitems, sum, sum_recv;
  int values[NITEMS], values_recv[NITEMS], values_recv_ref[NITEMS];
  MPI_Datatype contig_type;
  MPI_Op custom_sum;

  MPI_Init(&argc, &argv);
  MPI_Comm_size(MPI_COMM_WORLD, &nproc);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);

  //Set up periodic domain
  left = rank - 1;
  if (left < 0) left = nproc - 1;
  right = rank + 1;
  if (right > nproc - 1) right = 0;

  for (iitems = 0; iitems < NITEMS; ++iitems){
    values[iitems] = rank;
  }

  MPI_Type_contiguous(NITEMS, MPI_INT, &contig_type);
  MPI_Type_commit(&contig_type);

  //Create an operator to add together this type
  MPI_Op_create(&adder, 1, &custom_sum);

  MPI_Reduce(values, values_recv, 1, contig_type, custom_sum, 0, 
      MPI_COMM_WORLD);
  MPI_Reduce(values, values_recv_ref, NITEMS, MPI_INT, MPI_SUM, 0,
      MPI_COMM_WORLD);

  MPI_Type_free(&contig_type);

  sum = 0;
  sum_recv = 0;
  for (iitems = 0; iitems < NITEMS; ++iitems){
    sum += values_recv[iitems];
    sum_recv += values_recv[iitems];
  }

  if (rank == 0) {
    printf("Sum over all processors = %5d\n", sum);
    printf("Sum over all processors (reference version) = %5d\n", sum_recv);
  }

  MPI_Finalize();

}
